File: frontend/src/lib/revenueBoard.ts
// Client for the Total Revenue board.
//
// Mirrors backend/app/schemas/revenue_board.py. Shaped like lib/incomeBoard.ts on
// purpose — same axios instance, same repeated-key serialisation, same rule about NULL.
//
// `number | null` IS NOT `number | 0`, and the distinction survives all the way to the
// screen. A null incentive means a deal matched but pays on something the statement does
// not print, so nothing is claimed; rendering it as ₹0 tells the reader "you are owed
// nothing" when the truth is "we could not confirm what you are owed". Every money field
// below that can be null is typed null, and every tile that shows one renders an em dash
// with the affected row count beside it.
import api from "@/lib/api";

export type RevenueScope = "auto" | "mine" | "agency";
export type RevenueBasis = "issue" | "travel";

/** The six statement types whose gross IS the agency's sale, in nav order.
 *  Byte-identical to models/income_board.SALE_SOURCES — they index `by_source`. */
export const SALE_SOURCES = [
  "bsp", "ndc", "lcc-detailed", "tp-gds", "tp-lcc", "tp-api",
] as const;
export type SaleSource = (typeof SALE_SOURCES)[number];

/** Short labels for a column header, where the registry's sheet titles are too long. */
export const SOURCE_SHORT: Record<string, string> = {
  bsp: "BSP",
  ndc: "NDC",
  "lcc-detailed": "LCC",
  "tp-gds": "TP GDS",
  "tp-lcc": "TP LCC",
  "tp-api": "TP API",
};

export interface RevenueTotals {
  sale: number | null;
  sale_rows: number;
  incentive: number | null;
  iata_commission: number | null;
  unpriced_rows: number;
  needs_data_rows: number;
  unmatched_rows: number;
  unattributed_airline_rows: number;
  /** Third Party API rows. No carrier BY NATURE, which is not the same claim as a
   *  carrier that could not be resolved — they never share a bucket. */
  not_carrier_attributed_rows: number;
  undated_rows: number;
  not_counted_rows: number;
  /** What the board left out of `sale` on purpose, so the gap has an answer on screen. */
  not_counted_sale: number | null;
}

export interface SourcePoint {
  source: string;
  category: string;
  label: string;
  sale: number | null;
  rows: number;
  incentive: number | null;
  priced_rows: number;
  unpriced_rows: number;
  not_counted_rows: number;
  share_pct: number;
}

export interface CategoryPoint {
  category: string;
  sale: number | null;
  rows: number;
  incentive: number | null;
  share_pct: number;
  sources: string[];
}

export interface MonthPoint {
  ym: string;
  label: string;
  sale: number | null;
  incentive: number | null;
  rows: number;
  by_source: Record<string, number>;
}

export interface AirlineRevenuePoint {
  airline_id: number | null;
  airline: string;
  sale: number | null;
  incentive: number | null;
  rows: number;
  share_pct: number;
  by_source: Record<string, number>;
  by_incentive_type: Record<string, number>;
}

export interface SupplierPoint {
  supplier_id: number | null;
  supplier: string;
  supplier_code: string | null;
  branch: string | null;
  sale: number | null;
  incentive: number | null;
  rows: number;
  share_pct: number;
}

export interface ProductPoint {
  product: string | null;
  sale: number | null;
  rows: number;
  share_pct: number;
}

export interface IncentiveTypePoint {
  incentive_type: string;
  amount: number | null;
  rows: number;
  share_pct: number;
}

export interface LinkedStatementPoint {
  source: string;
  label: string;
  category: string;
  rows: number;
  /** What the figure IS — "Flown revenue", "Debit memos". Never the word sale. */
  measure_label: string;
  measure: number | null;
  schema_unverified: boolean;
}

export interface NotCountedPoint {
  /** The Report Download's own NET_* wording, so the same row gives the same answer in
   *  the downloadable workbook. Never a phrasing invented for this screen. */
  reason: string;
  rows: number;
  amount: number | null;
  sources: string[];
}

export interface RevenueSummary {
  scope: string;
  basis: string;
  currency: string | null;
  date_from: string | null;
  date_to: string | null;
  totals: RevenueTotals;
  by_category: CategoryPoint[];
  by_source: SourcePoint[];
  by_month: MonthPoint[];
  by_airline: AirlineRevenuePoint[];
  by_supplier: SupplierPoint[];
  by_product: ProductPoint[];
  not_counted: NotCountedPoint[];
  other_currencies: Record<string, number>;
}

export interface AirlineDetail {
  airline_id: number | null;
  airline: string;
  totals: RevenueTotals;
  by_source: SourcePoint[];
  by_month: MonthPoint[];
  by_incentive_type: IncentiveTypePoint[];
  linked: LinkedStatementPoint[];
}

export interface RevenueFilterOptions {
  airlines: { id: number | null; name: string }[];
  suppliers: { id: number | null; name: string; code: string | null }[];
  sources: string[];
  categories: string[];
  months: string[];
  currencies: string[];
  incentive_types: string[];
  can_view_agency: boolean;
  default_scope: string;
}

export interface RevenueQuery {
  scope?: RevenueScope;
  basis?: RevenueBasis;
  date_from?: string;
  date_to?: string;
  airline?: number[];
  supplier?: number[];
  source?: string[];
  category?: string[];
  currency?: string;
  top?: number;
}

// Repeated keys, not comma-joined: FastAPI reads `airline=1&airline=2` as a list and
// would take "1,2" as a single malformed int.
const LIST_PARAMS = { paramsSerializer: { indexes: null as null } };

export async function fetchRevenueSummary(q: RevenueQuery): Promise<RevenueSummary> {
  const { data } = await api.get<RevenueSummary>("/dashboard/revenue/summary", {
    params: q,
    ...LIST_PARAMS,
  });
  return data;
}

export async function fetchIncentiveTypes(q: RevenueQuery): Promise<IncentiveTypePoint[]> {
  const { data } = await api.get<IncentiveTypePoint[]>(
    "/dashboard/revenue/by-incentive-type",
    { params: q, ...LIST_PARAMS },
  );
  return data;
}

export async function fetchAirlineDetail(
  airlineId: number,
  q: RevenueQuery,
): Promise<AirlineDetail> {
  const { data } = await api.get<AirlineDetail>(
    `/dashboard/revenue/airline/${airlineId}`,
    { params: q, ...LIST_PARAMS },
  );
  return data;
}

export async function fetchLinkedStatements(
  q: { scope?: RevenueScope; airline?: number },
): Promise<LinkedStatementPoint[]> {
  const { data } = await api.get<LinkedStatementPoint[]>("/dashboard/revenue/linked", {
    params: q,
  });
  return data;
}

export async function fetchRevenueFilters(
  scope: RevenueScope,
): Promise<RevenueFilterOptions> {
  const { data } = await api.get<RevenueFilterOptions>("/dashboard/revenue/filters", {
    params: { scope },
  });
  return data;
}

/** The last 12 months, as the board's default window — the same default the income
 *  board uses, so switching tabs does not silently change the period. */
export function defaultRange(): { date_from: string; date_to: string } {
  const now = new Date();
  const to = new Date(now.getFullYear(), now.getMonth() + 1, 0);
  const from = new Date(now.getFullYear(), now.getMonth() - 11, 1);
  const iso = (d: Date) => d.toISOString().slice(0, 10);
  return { date_from: iso(from), date_to: iso(to) };
}

File: backend/app/schemas/revenue_board.py
"""Response shapes for the revenue board — what the loaded statements SOLD.

Every money field that can be NULL is typed `float | None`, and that is deliberate
rather than defensive, for the reason schemas/income_board.py gives: `incentive` is NULL
when a deal matched but pays on something the statement does not print, so nothing is
claimed, and a board that renders that as 0 has told the reader "you are owed nothing"
when the truth is "we could not confirm what you are owed".

THIS BOARD ADDS A THIRD STATE and it needs the same care. `unpriced_rows` counts lines
that carry sale and have never been through the commission engine. Their incentive is
also NULL, but for a reason that has nothing to do with the deal — the remedy is to
press Run, not to find a missing column — so the two are counted separately and shown
separately, never added.

AND A FOURTH FIGURE THAT IS NOT REVENUE. `not_counted_sale` is the value of the lines
this board deliberately leaves out of the total: a cancelled aggregator booking, an LCC
payment movement, an NDC ticket already settled through BSP, a statement's own footer
line. It is reported so that "the total is smaller than my statement's total" has an
answer on screen, and it is never added to `sale`.
"""
from __future__ import annotations

from pydantic import BaseModel


class RevenueTotals(BaseModel):
    """The KPI band."""
    # Gross as printed, signed, over the lines that count. Never the sum of every
    # statement: twelve types cover six businesses, and adding them all roughly
    # double-counts the BSP and LCC families.
    sale: float | None
    sale_rows: int
    # What those lines earned. NULL-safe by construction — see measures.incentive_sum.
    incentive: float | None
    # A separate entitlement, never folded into the incentive. The separation is
    # deliberate in services/commission/runner.py and models/bsp_statement.py.
    iata_commission: float | None

    # The honesty counters that must travel with the money.
    unpriced_rows: int
    needs_data_rows: int
    unmatched_rows: int
    # Lines whose carrier could not be resolved. Excludes Third Party API, which has no
    # carrier BY NATURE rather than by failure — see `not_carrier_attributed_rows`.
    unattributed_airline_rows: int
    not_carrier_attributed_rows: int
    # A date the vendor printed that no pattern could read. The row is real and its sale
    # counts; it simply has no month, and a blank month bucket must not read as no sales.
    undated_rows: int

    # Deliberately out of `sale`, and reported so the difference has an explanation.
    not_counted_rows: int
    not_counted_sale: float | None


class SourcePoint(BaseModel):
    """One statement type: BSP, NDC, LCC Detailed, or one of the three third-party ones."""
    source: str
    category: str          # BSP | LCC | Third Party
    label: str
    sale: float | None
    rows: int
    incentive: float | None
    priced_rows: int
    unpriced_rows: int
    not_counted_rows: int
    share_pct: float


class CategoryPoint(BaseModel):
    """BSP / LCC / Third Party, folded from `by_source` rather than grouped again."""
    category: str
    sale: float | None
    rows: int
    incentive: float | None
    share_pct: float
    sources: list[str]


class MonthPoint(BaseModel):
    ym: str
    label: str
    sale: float | None
    incentive: float | None
    rows: int
    #: source -> sale, for the stacked column. Absent keys are zero months, not gaps.
    by_source: dict[str, float]


class AirlineRevenuePoint(BaseModel):
    """One carrier, across every statement type it appears in.

    `airline_id` NULL is the unattributed bucket. `by_source` is what makes the row
    answer the question the board exists for — the same carrier's BSP, NDC and
    consolidator business side by side instead of on three screens.
    """
    airline_id: int | None
    airline: str
    sale: float | None
    incentive: float | None
    rows: int
    share_pct: float
    by_source: dict[str, float]
    by_incentive_type: dict[str, float]


class SupplierPoint(BaseModel):
    supplier_id: int | None
    supplier: str
    supplier_code: str | None
    branch: str | None
    sale: float | None
    incentive: float | None
    rows: int
    share_pct: float


class ProductPoint(BaseModel):
    """Third Party API only — the one multi-product source.

    An aggregator file carries hotel, flight, train, bus and car bookings side by side,
    and only the flight rows have a carrier at all. Reporting this source by product
    rather than by airline is what keeps a hotel night out of an airline's total.
    """
    product: str | None
    sale: float | None
    rows: int
    share_pct: float


class IncentiveTypePoint(BaseModel):
    """PLB, Super PLB, Transaction Fee, and the eight others.

    Read out of `incentive_breakdown` rather than from a fixed column list, so an
    incentive type added to a deal shows up without a code change — `incentive_type` is
    a String on `deal_incentives`, and the canonical eleven in api/v1/tickets.py are a
    convention, not a constraint.
    """
    incentive_type: str
    amount: float | None
    rows: int
    share_pct: float


class NotCountedPoint(BaseModel):
    """One reason a line was held out of the sale total, with what it was worth.

    The wording is `report_download/columns.py`'s own NET_* vocabulary, not a phrasing
    invented here, so the answer to "why is this smaller than my statement's total" is
    the same sentence on this board and in the downloadable workbook.
    """
    reason: str
    rows: int
    amount: float | None
    sources: list[str]


class RevenueSummaryResponse(BaseModel):
    scope: str
    basis: str
    currency: str | None
    date_from: str | None
    date_to: str | None
    totals: RevenueTotals
    by_category: list[CategoryPoint]
    by_source: list[SourcePoint]
    by_month: list[MonthPoint]
    by_airline: list[AirlineRevenuePoint]
    by_supplier: list[SupplierPoint]
    by_product: list[ProductPoint]
    #: Why the board's total is smaller than the sum of the statements, line by reason.
    not_counted: list[NotCountedPoint]
    #: Currencies present in scope beyond the one selected, with their row counts.
    #: Totals are never converted or added across currencies, so the board scopes to one
    #: and says what it left out rather than silently dropping it.
    other_currencies: dict[str, int]


class AirlineDetailResponse(BaseModel):
    """One carrier's whole picture — the drill-down behind a row of the matrix."""
    airline_id: int | None
    airline: str
    totals: RevenueTotals
    by_source: list[SourcePoint]
    by_month: list[MonthPoint]
    by_incentive_type: list[IncentiveTypePoint]
    #: The statements that restate this carrier's business rather than adding to it —
    #: TGQ HMPR, the memos, the LCC ledgers. Counts and their own measures, never summed
    #: into sale.
    linked: list[LinkedStatementPoint]


class LinkedStatementPoint(BaseModel):
    source: str
    label: str
    category: str
    rows: int
    #: The measure that type is actually about: ADM exposure, flown revenue, deposits.
    #: Named rather than called "sale", because it is not sale.
    measure_label: str
    measure: float | None
    #: True where the statement's layout was built without a real sample
    #: (`registry.schema_unverified`), so the figure carries that caveat on screen.
    schema_unverified: bool


class RevenueFilterOptions(BaseModel):
    airlines: list[dict]
    suppliers: list[dict]
    sources: list[str]
    categories: list[str]
    months: list[str]
    currencies: list[str]
    incentive_types: list[str]
    can_view_agency: bool
    default_scope: str


AirlineDetailResponse.model_rebuild()
